pkg/volume/csi/csi_attacher.go

```diff
@@ -106,22 +107,33 @@ func (c *csiAttacher) Attach(spec *volume.Spec, nodeName types.NodeName) (string
 		},
 	}
 
-	_, err = c.k8s.StorageV1().VolumeAttachments().Create(context.TODO(), attachment, metav1.CreateOptions{})
-	alreadyExist := false
+	// Begin watching for this volume attachment prior to creating it to avoid race conditions.
+	if _, err = c.plugin.volumeAttachmentInformerHook.CreateWatch(attachID, watchChannelBufferSize); err != nil {
+		return "", errors.New(log("failed to create watch with name: %s, %v", attachID, err))
+	}
+	defer c.plugin.volumeAttachmentInformerHook.DeleteWatch(attachID)
+
+	alreadyExist := true
+	_, err = c.plugin.volumeAttachmentLister.Get(attachID)
 	if err != nil {
-		if !apierrors.IsAlreadyExists(err) {
-			return "", errors.New(log("attacher.Attach failed: %v", err))
+		if !apierrors.IsNotFound(err) {
+			return "", errors.New(log("failed to get volume attachment from lister: %v", err))
 		}
-		alreadyExist = true
+		alreadyExist = false
```

Comment:
just a nit: alreadyExist seems not very necessary here.

If err != nil || va == nil {
    return with error
}
// continue since err== nil and va != nil

Reply:
do you want to simplify the logic here if possible?

Reply:
Yes, I will get around to this. I'm currently looking at running a little test to see at what point do we successfully verify an attachment has succeeded, because we either successfully verify this at the moment we fetch it from the lister, otherwise we verify it by waiting for an event. If it seems like the majority of the time we successfully verify it the moment we fetch from the lister, then it may be worth removing the watch logic entirely.